src/service/fixed-layer.js

```diff
@@ -0,0 +1,417 @@
+   * whether they are currently visible and fixed.
+   * @return {!Promise}
+   */
+  update() {
+    if (this.fixedElements_.length == 0) {
+      return Promise.resolve();
+    }
+
+    // Some of the elements may no longer be in DOM.
+    /** @type {!Array<number>} */
+    const toRemove = [];
+    this.fixedElements_.forEach((fe, i) => {
+      if (!this.doc.contains(fe.element)) {
+        toRemove.push(i);
+      }
+    });
+    if (toRemove.length > 0) {
+      toRemove.forEach(i => {
+        this.fixedElements_.splice(i, 1);
+      });
+    }
+
+    /** @type {!Object<string, !FixedElementStateDef>} */
+    const fixedMap = {};
+    let hasTransferables = false;
+    return this.vsync_.runPromise({
+      measure: () => {
+        this.fixedElements_.forEach(fe => {
+          const element = fe.element;
+          const styles = this.doc.defaultView.getComputedStyle(element, null);
+          const position = styles.getPropertyValue('position');
+          const top = styles.getPropertyValue('top');
+          const bottom = styles.getPropertyValue('bottom');
+          const isFixed = (
+              position == 'fixed' &&
+              element.offsetWidth > 0 &&
+              element.offsetHeight > 0);
+          const isTransferrable = (
+              isFixed &&
+              (this.isAllowedCoord_(top) || this.isAllowedCoord_(bottom)));
+          if (isTransferrable) {
+            hasTransferables = true;
+          }
+          fixedMap[fe.id] = {
+            fixed: isFixed,
+            transferrable: isTransferrable,
+            top: styles.getPropertyValue('top'),
+            zIndex: styles.getPropertyValue('z-index'),
+          };
+        });
+      },
+      mutate: () => {
+        if (hasTransferables && this.transfer_) {
+          const fixedLayer = this.getFixedLayer_();
+          if (fixedLayer.className != this.doc.body.className) {
+            fixedLayer.className = this.doc.body.className;
+          }
+        }
+        this.fixedElements_.forEach((fe, i) => {
+          const state = fixedMap[fe.id];
+          if (state) {
+            this.mutateFixedElement_(fe, i, state);
+          }
+        });
+      }
+    }).catch(error => {
+      // Fail silently.
+      setTimeout(() => {throw error;});
+    });
+  }
+
+  /**
+   * We currently only allow elements with `top: 0` or `bottom: 0`.
+   * @param {string} s
+   * @return {boolean}
+   */
+  isAllowedCoord_(s) {
+    return (s && parseInt(s, 10) == 0);
+  }
+
+  /**
+   * @param {!Element} element
+   * @param {string} selector
+   * @param {!Object<string, !FixedElementDef>} idMap
+   * @private
+   */
+  setupFixedElement_(element, selector, idMap) {
+    let fixedId = element.getAttribute('i-amp-fixedid');
+    let fe;
+    if (fixedId) {
+      // Already seen.
+      fe = idMap[fixedId];
+      fe.selectors.push(selector);
+    } else {
+      fixedId = 'F' + this.fixedElements_.length;
```

Comment:
Switched to counter. Although, this was ok as well - these are calculated only once. But still a good idea.